stdlib/public/core/NFC.swift

```diff
@@ -0,0 +1,201 @@
+          // We did not find a composition between the two. If our current class
+          // is 0, then set that as the new composee and return whatever built
+          // up scalar we have. Otherwise, add our current scalar to the buffer
+          // for eventually removal!
+          
+          guard currentCCC == 0 else {
+            buffer.append(current)
+            continue
+          }
+          
+          composee = current.scalar
+          return l
+        }
+        
+        // We found a composition! Record it as our new composee and repeat the
+        // process.
+        composee = p
+        continue
+      }
+      
+      // We only care about the last's ccc.
+      let lastCCC = lastNormData >> 3
+      
+      // Check if our current scalar is not blocked from our current composee.
+      // In this case blocked means there is some scalar whose class (lastClass)
+      // is either == 0 or >= currentClass.
+      //
+      // Example:
+      //
+      //     "z\u{0335}\u{0327}\u{0324}\u{0301}"
+      //
+      // In this example, there are several combining marks following a 'z', but
+      // none of them actually compose with the composee 'z'. However, the last
+      // scalar U+0301 does actually compose. So this check makes sure that the
+      // last scalar doesn't have any scalar in between it and the composee that
+      // would otherwise "block" it from composing.
+      guard lastCCC < currentCCC else {
+        // We had a scalar block it. That means our current scalar is either a
+        // starter or has a same class (preserve ordering).
+        
+        guard currentCCC == 0 else {
+          // Not a starter, stick it at the end of the buffer and keep going!
+          
+          buffer.append(current)
+          continue
+        }
+        
+        // Starters are the "start" of a new normalization segment. Set it as
+        // the new composee and return our current composee. This will trigger
+        // any other scalars in the buffer to be emitted before we handle
+        // composing this new composee.
+        composee = current.scalar
+        return l
+      }
+      
+      // There were no blockers! Attempt to compose the two! (Again, if our rhs
+      // scalar IS NFC_QC, then it can never compose with anything previous to
+      // it).
+      guard !currentIsNFCQC, let p = compose(l, and: current.scalar) else {
+        // No composition found. Stick it at the end of the buffer with the rest
+        // of non-composed scalars.
+        
+        buffer.append(current)
+        continue
+      }
+      
+      // They composed! Assign the composition as our new composee and iterate
+      // to the next scalar.
+      composee = p
+    }
+    
+    // If we have a leftover composee, make sure to return it.
+    return composee.take()
+  }
+}
+
+extension Unicode.NFC: Sequence {
+  internal func makeIterator() -> Iterator {
+    Iterator(iterator: base.nfd.makeIterator())
```

Comment:
Is it perhaps worth reserving capacity in the buffer? Otherwise it will be the empty array singleton, and the first append will allocate IIRC.

Reply:
So, I experimented with this actually, and what I found was that reserving capacity for all strings didn't make sense. The reason for this is that a lot of the common strings don't even go use the buffer at all, so reserving capacity for this buffer made some of the benchmarks slower for things like ascii, latiny strings, etc. For strings that do use the buffer, this provided little to no difference in terms of benchmarking performance. I only ever saw negligible wins with our zalgo benchmarks when I did something crazy like `reserveCapacity(100)`.